### DISCUSSION TITLE
Fix farm orientation(north up) and plot turbine with fi

### DISCUSSION THREAD

**Discussion by dhcho347** [NatLabRockies/floris#530] [Show and tell]
1. Fix Farm Orientation :
 a. before change

![0before_chg](https://user-images.githubusercontent.com/70368625/201642126-34d593c9-9b6b-4ec0-8317-322745b99073.png)

 b. after change (without planar flowfield broaden)
![after_chg_45deg_without broaden flowfield](https://user-images.githubusercontent.com/70368625/201642316-ad2aa433-4b6b-43cf-910e-e19f1daa1908.png)

 c. after change with flowfield broaden (0,90,180,270 deg)
![after_chg_0deg](https://user-images.githubusercontent.com/70368625/201642392-60d7b676-7018-42c8-92b0-5aca79338766.png)
 (0,90,180,270 deg)+45deg
![after_chg_45deg](https://user-images.githubusercontent.com/70368625/201642414-6a20881f-d203-45af-bc56-68175587c098.png)

 d. change things
![chgne_things](https://user-images.githubusercontent.com/70368625/201651132-48f39223-1947-47aa-90c3-a721fce69a0e.png)

 e. myCode
 - Repo: https://github.com/dhcho347/floris/tree/fix_farm_orientation
 -  ex: https://github.com/dhcho347/floris/blob/fix_farm_orientation/examples/vis_03_horizontal_plane.txt
 - I wanted to minimize code changes. I didn't add a new function, just using an existing one with small change
 - Change things:
   v grid.py: class flowfiledplanar Grid method set_grid (Braoden flowfield, rotate mesh grid)
   v utilities.py: minor change to ftn rotate_coordinates_rel_west()
   v floris_interface.py: method get_plane_of_points() (reverse rotating to north up)
  
 f. refer: #348, #357, #372

2. plot with fi
 a. before change yaw angle (with fi.reinitialize)
![fi_before](https://user-images.githubusercontent.com/70368625/201644956-81255acb-5fb9-4a43-af06-f4fb4efd5ee1.png)

 b. after change (with fi.reinitialize)
![fi_after_chg](https://user-images.githubusercontent.com/70368625/201645134-7fca8ccb-3475-466e-84ce-f7f1ea173b6e.png)

 c. after change. different WTG types
![fi_after_chg_diffWTG](https://user-images.githubusercontent.com/70368625/201645798-929ac317-a69a-4c08-b367-c2aa6a676455.png)

 d. myCode
 - Repo: https://github.com/dhcho347/floris/tree/plot_turbine_with_fi
 -  ex: https://github.com/dhcho347/floris/blob/plot_turbine_with_fi/examples/vis_03_horizontal_plane.py
  - change things
   v visualization.py: ftn plot_turbines
   v fi update before fi.calculate_horizontal_plane() to maintain cal results to use texting

 e. refer: #448, #451, #455, #512

3. Github
- This is my first time using Github, the form and usage may not be correct. It's not intentional, so let me know and I'll fix it.
- Should copied floris repo be private? It's currently public, let me know, I'll change it
- Thanks to NREL team for sharing and making these tool

**Comment by rafmudaf** [comment]
Hi @dhcho347 Thanks for making these changes and sharing! Could you open a draft pull request to the develop branch so that we can start to understand the code changes?

**Reply by dhcho347** [reply]
Hi @rafmudaf , Okay I'll try after reading how to pull request. Thanks

**Reply by paulf81** [reply]
Wanted to link now to pull request #533

**Comment by MYMahfouz** [comment]
Hi,

Thanks for all the work done in rotating back the wind farm for plots.

I have one minor comment. This method only works if the center of rotations are small or zero. If the center of rotation increases it will cause the turbines locations in the farm to be inaccurate with the current approach. I suggest updating the rotate_coordinates_rel_west to the following :

```python
def rotate_coordinates_rel_west(wind_directions, coordinates, inv_rot=False,layout_x=0,layout_y=0):
    # Calculate the difference in given wind direction from 270 / West
    wind_deviation_from_west = wind_delta(wind_directions)
    wind_deviation_from_west = np.reshape(wind_deviation_from_west, (len(wind_directions), 1, 1))
      

    # Construct the arrays storing the turbine locations
    # x_coordinates, y_coordinates, z_coordinates = coordinates.T
    if isinstance(coordinates, (np.ndarray, np.generic) ) :
        x_coordinates, y_coordinates, z_coordinates = coordinates.T
    else :
        x_coordinates, y_coordinates, z_coordinates = coordinates #dh. to handle additional input type (mesh grid)

    if inv_rot : 
        wind_deviation_from_west = -wind_deviation_from_west #dh. for inverse rotation
        x_center_of_rotation = (np.min(layout_x) + np.max(layout_x)) / 2
        y_center_of_rotation = (np.min(layout_y) + np.max(layout_y)) / 2
    else:
        # Find center of rotation - this is the center of box bounding all of the turbines
        x_center_of_rotation = (np.min(x_coordinates) + np.max(x_coordinates)) / 2
        y_center_of_rotation = (np.min(y_coordinates) + np.max(y_coordinates)) / 2
    

    # Rotate turbine coordinates about the center
    x_coord_offset = x_coordinates - x_center_of_rotation
    y_coord_offset = y_coordinates - y_center_of_rotation
    

    
    if inv_rot : 
        x_coord_rotated = (
            x_coord_offset * cosd(wind_deviation_from_west)
            - y_coord_offset * sind(wind_deviation_from_west)
        )
        # pdb.set_trace()
        x_coord_rotated =x_coord_rotated + x_center_of_rotation
        y_coord_rotated = (
            x_coord_offset * sind(wind_deviation_from_west)
            + y_coord_offset * cosd(wind_deviation_from_west)
        )
        y_coord_rotated =y_coord_rotated + y_center_of_rotation
    else:
        x_coord_rotated = (
            x_coord_offset * cosd(wind_deviation_from_west)
            - y_coord_offset * sind(wind_deviation_from_west)
            + x_center_of_rotation
        )
        y_coord_rotated = (
            x_coord_offset * sind(wind_deviation_from_west)
            + y_coord_offset * cosd(wind_deviation_from_west)
            + y_center_of_rotation
        )
    
    z_coord_rotated = np.ones_like(wind_deviation_from_west) * z_coordinates
    return x_coord_rotated, y_coord_rotated, z_coord_rotated
```


Also updating the line calling the  rotate_coordinates_rel_west function within the get_plane_of_points function to the following:

```python
x_flat2, y_flat2, z_flat2 = rotate_coordinates_rel_west(self.floris.flow_field.wind_directions, (x_flat, y_flat, z_flat), inv_rot=True ,layout_x=self.floris.farm.layout_x,layout_y=self.floris.farm.layout_y)
```

**Reply by dhcho347** [reply]
Hi, @MYMahfouz,
Thank you for checking my codes and a great enhancement.

You are right. 
Problems arise when the center of the layout and the center of the flowfield do not match.

for setting bounds of planarflowfield in grid.py
https://github.com/NREL/floris/blob/6124d2a82a7a823722210bc2e8516d355ba19eb3/floris/simulation/grid.py#L363-L364

1. with symmetric expansion  (with both 10, 10) 
https://github.com/NREL/floris/blob/ed93d54a2c57670de3cd838a824e327b7a1e7b36/floris/simulation/grid.py#L347
  => **It's okay**   [center of turbines = (1500, 400) and flowfield = (1500,400) ]

2. with asymmetric flowfield expansion  (with  5 and 10) 
![image](https://user-images.githubusercontent.com/70368625/232974049-68a435fb-3cd0-437e-9552-b558c2f7ace7.png)
=> **It's wrong**  [center of turbines = (1500, 400) and flowfield = (1815,400) ]

![Figure_1_with my](https://user-images.githubusercontent.com/70368625/232969861-f1da0115-596e-4ff0-9f2b-e8598e13ba9c.png)
If the two centers do not coincide, the turbines are positioned incorrectly, as shown in the figure above.

I update my code with yours except the third if-else statement
(the first section and the second section look the same.
l think we don't need to divide it)

I missed it because I only tested it with the symmetric expansion examples.

Thank you for corrections!

**Reply by Bartdoekemeijer** [reply]
FYI: PR #578 should also allow you to plot flow fields in the internal frame of reference (without the rotation).

**Reply by dhcho347** [reply]
Hi, @Bartdoekemeijer 
Thank you for the notice. I will also check and learn it.

**Comment by Bartdoekemeijer** [comment]
Hi @MYMahfouz. Thanks for doing this, that is super helpful! Could you open a [pull request](https://github.com/NREL/floris/pulls) to get your work merged in to the _develop_ branch?

**Reply by MYMahfouz** [reply]
I have created this pull request. I am new to pull requests so I hope i did everything properly. 
pull request https://github.com/NREL/floris/pull/556

**Comment by RodAc0s** [comment]
Hi, 
Sorry is there any specific reason why the power per turbine increases on free flow after the yaw command is applied to one of the turbines (an increase of 7.22 kW per turbine in this case)? When the upstream wind has no interaction with the wakes the power should be the same for the non-yawed turbines, as it is shown on the code that is not using yaw misalignment, but with this function, it shows an increase in the powers.


![180°](https://user-images.githubusercontent.com/122547278/212112819-845eb32e-c061-4556-90d6-4ab39c6e7e23.PNG)

**Reply by rafmudaf** [reply]
@RodAc0s Is your question specific to the topic discussed here or is it a general question? If this is a general question, please open a new discussion.

**Reply by dhcho347** [reply]
Hi, @RodAc0s 
Sorry, I couldn't reproduce the same situation.

# Jensen
![Figure_Jensen_0](https://user-images.githubusercontent.com/70368625/233277609-aae7270c-1056-4e64-9a78-927ba1151349.png)


# GCH
![Figure_gch_0](https://user-images.githubusercontent.com/70368625/233277618-7f37de97-2cc1-449f-968b-304c928b388b.png)

**Comment by rafmudaf** [comment]
Hi @dhcho347 I believe this discussion has been resolved ultimately by #578 and #663. Can you confirm that this is true? If so, we can close this issue and your open pull request #533. By the way, thank you very much for documenting this so well and creating the pull request with a suggested fix. It was very helpful in arriving at the final implementation.

**Reply by rafmudaf** [reply]
I'm closing this discussion, but please reopen or update if I've left anything out. @dhcho347 Thanks again for your efforts, and for documenting this approach so well both in this discussion and your corresponding pull request.

**Reply by dhcho347** [reply]
Sorry for the late reply.
Glad you came up with a final solution. I will also use the latest version.
